src/main/scala/lms/transformation/tensor.scala

```diff
@@ -27,7 +27,10 @@ import Backend._
  * the typeless frontend is more friendly to IR transformations.
  * This is similar to how MLIR is typeless and has been easy to run transformations.
  */
-trait FixedSizeTensorFrontEnd extends Base with PrimitiveOps with ArrayOps {
+object FixedSizeTensorTypeLess extends Base with PrimitiveOps with ArrayOps {
```

Comment:
this extends Base ... cannot be removed, because in this typeless frontend, I actually used typed frontends :(  
not nice but this is with minimal effort